src/psyGen.py

```diff
@@ -701,6 +664,182 @@ def gen_code(self, parent):
+                remote_name += start_postfix
+            graph.edge(local_name, remote_name, color="green")
+        elif self.parent:
+            remote_name = self.parent.dag_name + end_postfix
+            graph.edge(local_name, remote_name, color="blue")
+        node = self.backward_dependence()
+        local_name = self.dag_name
+        if self.children:
+            local_name += start_postfix
+        if node:
+            remote_name = node.dag_name
+            if node.children:
+                remote_name += end_postfix
+            graph.edge(remote_name, local_name, color="red")
+        elif self.parent:
+            remote_name = self.parent.dag_name + start_postfix
+            graph.edge(remote_name, local_name, color="blue")
+        for child in self.children:
+            child.dag_gen(graph)
+
+    @property
+    def dag_name(self):
+        ''' return the base dag name for this node '''
+        return "node_" + str(self.abs_position)
+
+    @property
+    def args(self):
+        '''Return the list of arguments associated with this node. The default
+        implementation assumes the node has no directly associated
+        arguments (i.e. is not a Call class or subclass). Arguments of
+        any of this nodes descendents are considered to be
+        associated. '''
+        args = []
+        for call in self.calls():
+            args.extend(call.args)
+        return args
+
+    def backward_dependence(self):
+        '''Returns the closest preceding Node that this Node has a direct
+        dependence with or None if there is not one. Only Nodes with
+        the same parent as self are returned. Nodes inherit their
+        descendents dependencies.'''
+        dependence = None
+        for arg in self.args:
+            dependent_arg = arg.backward_dependence()
+            if dependent_arg:
+                node = dependent_arg.call
+                while node.depth > self.depth:
+                    node = node.parent
+                if self.sameParent(node):
+                    if not dependence:
+                        dependence = node
+                    else:
+                        if dependence.position < node.position:
+                            dependence = node
+        return dependence
+
+    def forward_dependence(self):
+        '''Returns the closest following Node that this Node has a direct
+        dependence with or None if there is not one. Only Nodes with
+        the same parent as self are returned. Nodes inherit their
+        descendents dependencies.'''
+        dependence = None
+        for arg in self.args:
+            dependent_arg = arg.forward_dependence()
+            if dependent_arg:
+                node = dependent_arg.call
+                while node.depth > self.depth:
+                    node = node.parent
+                if self.sameParent(node):
+                    if not dependence:
+                        dependence = node
```

Comment:
I believe you understand the logic now but for the record we are looking for the closest dependence. There may be more than one and the closest is the one that constrains us. Therefore we do not return when we find the first dependence, we search through all of them.